src/main/java/org/checkerframework/specimin/unsolved/FullyQualifiedNameGenerator.java

```diff
@@ -376,21 +378,61 @@ else if (expr.isCastExpr()) {
 
     // Handle the cases where the type of the expression can be inferred from surrounding context
     if (expr.hasParentNode()) {
-      @Nullable FullyQualifiedNameSet fromLHS = getFQNsFromSurroundingContextType(expr);
+      @Nullable FullyQualifiedNameSet fromLHS = getFQNsFromSurroundingContextType(expr, canRecurse);
       if (fromLHS != null) {
         return fromLHS;
       }
     }
 
+    // Handle binary expressions after surrounding context because the binary expression could be on
+    // the
+    // right-hand side where the left side type is known.
+    if (expr.isBinaryExpr()) {
+      BinaryExpr binary = expr.asBinaryExpr();
+      Operator operator = binary.getOperator();
+
+      // Boolean
+      if (operator == Operator.AND
+          || operator == Operator.OR
+          || operator == Operator.EQUALS
+          || operator == Operator.NOT_EQUALS
+          || operator == Operator.LESS
+          || operator == Operator.GREATER
+          || operator == Operator.LESS_EQUALS
+          || operator == Operator.GREATER_EQUALS) {
+        return new FullyQualifiedNameSet("boolean");
+      } else {
+        // Treat all other cases; type on one side is equal to the other
+        FullyQualifiedNameSet leftType = getFQNsForExpressionTypeImpl(binary.getLeft(), canRecurse);
+        FullyQualifiedNameSet rightType =
+            getFQNsForExpressionTypeImpl(binary.getRight(), canRecurse);
+
+        // Remaining operators only work with primitive/String types
+        // Safe to call isJavaLangOrPrimitiveName since any non-primitive/String type would be
+        // a synthetic type here
+        if (leftType.erasedFqns().size() == 1
+            && JavaLangUtils.isJavaLangOrPrimitiveName(leftType.erasedFqns().iterator().next())) {
+          return leftType;
+        }
+        if (rightType.erasedFqns().size() == 1
+            && JavaLangUtils.isJavaLangOrPrimitiveName(rightType.erasedFqns().iterator().next())) {
+          return rightType;
+        }
+
+        // TODO: alternates here? cases here are all primitives (except for +, which could be
+        // String)
```

Comment:
It's easy to cover all of the alternates here (we know the list of primitives), so we should just do so to avoid a headache later.

Reply:
> I can do this after I address the other failing cases, though.

It's not a priority, I agree. Leaving the TODO here is fine: we can come back to it or not later.